Extension/src/Debugger/configurationProvider.ts

```diff
@@ -71,67 +80,149 @@ export class QuickPickConfigurationProvider implements vscode.DebugConfiguration
-        return this.underlyingProvider.resolveDebugConfiguration ? this.underlyingProvider.resolveDebugConfiguration(folder, config, token) : undefined;
+        if (!config || !config.type) {
+            // When DebugConfigurationProviderTriggerKind is Dynamic, this function will be called with an empty config.
+            // Providing debug configs, and debugging should be called manually.
+            this.provideDebugConfigurations(folder).then(async configs => {
+                if (!configs || configs.length === 0) {
+                    Telemetry.logDebuggerEvent(DebuggerEvent.debugPanel, { "debugType": "debug", "folderMode": folder ? "folder" : "singleMode", "cancelled": "true" });
+                    return undefined; // aborts debugging silently
+                } else {
+                    await this.startDebugging(folder, configs[0], DebuggerEvent.debugPanel);
+                    return configs[0];
+                }
+            });
+        } else {
+            // resolveDebugConfigurationWithSubstitutedVariables will be called automatically after this.
+            if (config.type === 'cppvsdbg') {
+                // Handle legacy 'externalConsole' bool and convert to console: "externalTerminal"
+                if (config.hasOwnProperty("externalConsole")) {
+                    logger.getOutputChannelLogger().showWarningMessage(localize("debugger.deprecated.config", "The key '{0}' is deprecated. Please use '{1}' instead.", "externalConsole", "console"));
+                    if (config.externalConsole && !config.console) {
+                        config.console = "externalTerminal";
+                    }
+                    delete config.externalConsole;
+                }
+
+                // Fail if cppvsdbg type is running on non-Windows
+                if (os.platform() !== 'win32') {
+                    logger.getOutputChannelLogger().showWarningMessage(localize("debugger.not.available", "Debugger of type: '{0}' is only available on Windows. Use type: '{1}' on the current OS platform.", "cppvsdbg", "cppdbg"));
+                    return undefined; // Stop debugging
+                }
+            }
+            return config;
+        }
     }
 
+    /**
+     * This hook is directly called after 'resolveDebugConfiguration' but with all variables substituted.
+     * This is also ran after the tasks.json has completed.
+     *
+	 * Try to add all missing attributes to the debug configuration being launched.
+	 */
     resolveDebugConfigurationWithSubstitutedVariables(folder: vscode.WorkspaceFolder | undefined, config: vscode.DebugConfiguration, token?: vscode.CancellationToken): vscode.ProviderResult<vscode.DebugConfiguration> {
-        return this.underlyingProvider.resolveDebugConfigurationWithSubstitutedVariables ? this.underlyingProvider.resolveDebugConfigurationWithSubstitutedVariables(folder, config, token) : undefined;
-    }
-}
+        if (!config || !config.type) {
+            return undefined;
```

Comment:
Because we got bit by it, I think we should document that returning `null` here creates a launch.json and `undefined` is what we want.

Reply:
This was not addressed yet.